src/http_server/api/v1/metrics.c

```diff
@@ -299,19 +299,26 @@ void cb_metrics_prometheus(mk_request_t *request, void *data)
     qsort(metrics_arr, num_metrics, sizeof(char *), string_cmp);
 
     /* When a new metric starts add TYPE annotation. */
+    sds = flb_sds_cat(sds, "# HELP ", 7);
+    sds = flb_sds_cat(sds, metrics_arr[0], extract_metric_name_end_position(metrics_arr[0]));
+    sds = flb_sds_cat(sds, " fluentbit_metrics\n", 20);
     sds = flb_sds_cat(sds, "# TYPE ", 7);
     sds = flb_sds_cat(sds, metrics_arr[0], extract_metric_name_end_position(metrics_arr[0]));
     sds = flb_sds_cat(sds, " counter\n", 9);
```

Comment:
flb_sds_cat can lead to a memory allocation: https://github.com/fluent/fluent-bit/blob/master/src/flb_sds.c#L126

I think the reason that none of this code checks for an allocation failure is that the SDS string is initialized to length 1024 on line 207. 

However, with a few inputs and outputs configured, I was easily able to get Fluent Bit to output prom metrics that are longer than 1024 bytes.

I think the best thing to do would be to check for allocation failures on all of these lines. It'll be long and look ugly... but that's C programming...

You could also create a macro for just for this since there are so many lines where the same check is needed.

Reply:
I added an increase of sds in case of less than 200 bytes available for the next metric. Didn
't catched the memory problem, although it exceeds the 1024 bytes still in default configuration.

Reply:
@novegit Thanks for working on this!

That's not quite what I was thinking though...

Check out the code for `flb_sds_cat`: https://github.com/fluent/fluent-bit/blob/master/src/flb_sds.c#L118

It calls `flb_sds_avail` and `flb_sds_increase`. The issue is not that the string won't have enough memory when `flb_sds_cat` is called. The issue is that to be super safe, we always check if memory allocations succeed. As I understand it, in practice a memory allocation should essentially never fail on modern systems, but it's still a best practice to check.

The existing check on line 209 is what I am talking about.

For your code, something like this (conveniently there's an error label that does all the clean up):

```
sds = flb_sds_cat(sds, "# HELP ", 7);
if (!sds) {
     goto error;
}
sds = flb_sds_cat(sds, metrics_arr[i+1], extract_metric_name_end_position(metrics_arr[i+1]));
if (!sds) {
     goto error;
}
sds = flb_sds_cat(sds, " fluentbit_metrics\n", 20);
if (!sds) {
     goto error;
}
```

Reply:
@PettitWesley
thank you so much for your advise. Most of my C progamming skill is faded away since my last C project before 25years. Nowadays I am only a python programmer ;-)

When the memory allocating check is ok, I will try to improve the help text.